core/src/block_creation_loop.rs

```diff
@@ -238,40 +247,75 @@ fn start_loop(config: BlockCreationLoopConfig) {
             );
         }
 
-        // Wait for the voting loop to notify us, draining all pending messages and keeping the highest slot
+        // Race between parent ready notification and optimistic parent events.
+        // Parent ready is checked first and has priority if both channels are ready.
+        let window_source = select_biased! {
+            recv(ctx.leader_window_info_receiver) -> msg => {
+                msg.ok().map(ParentSource::ParentReady)
+            },
+            recv(optimistic_parent_receiver) -> msg => {
+                msg.ok().map(ParentSource::OptimisticParent)
+            },
+            default(Duration::from_secs(1)) => None,
+        };
+
+        // Drain any additional pending messages and keep the latest one
+        let window_source = match window_source {
+            Some(ParentSource::ParentReady(first)) => {
+                let latest = ctx
+                    .leader_window_info_receiver
+                    .try_iter()
+                    .last()
+                    .unwrap_or(first);
+                ParentSource::ParentReady(latest)
+            }
+            Some(ParentSource::OptimisticParent(first)) => {
+                let latest = optimistic_parent_receiver
+                    .try_iter()
+                    .last()
+                    .unwrap_or(first);
+                ParentSource::OptimisticParent(latest)
+            }
+            None => continue,
+        };
+
         let LeaderWindowInfo {
             start_slot,
             end_slot,
             // TODO: handle duplicate blocks by using the hash here
-            parent_block: (parent_slot, _),
+            parent_block: (parent_slot, parent_hash),
             skip_timer,
-        } = {
-            // Drain all pending messages and keep the latest one
-            let Some(info) = ctx
-                .leader_window_info_receiver
-                .recv_timeout(Duration::from_secs(1))
-                .ok()
-                .and_then(|window| {
-                    ctx.leader_window_info_receiver
-                        .try_iter()
-                        .last()
-                        .or(Some(window))
-                })
-            else {
-                continue;
-            };
-
-            info
+        } = match window_source {
```

Comment:
We should just combine this with the previous match and add the trace logs there